src/hdmf/container.py

```diff
@@ -753,12 +753,23 @@ def _generate_field_html(self, key, value, level, access_code):
     def _generate_array_html(self, array, level):
         """Generates HTML for array data"""
 
-        read_io = self.get_read_io()  # if the Container was read from file, get IO object
-        if read_io is not None: # Note that sometimes numpy array have a read_io attribute
-            repr_html = read_io.generate_dataset_html(array)
-        else:
+        is_numpy_array = isinstance(array, np.ndarray)
+        it_was_read_with_io = self.get_read_io() is not None
+        is_data_io = isinstance(array, DataIO)
+
+        if is_numpy_array:
             array_info_dict = get_basic_array_info(array)
             repr_html = generate_array_html_repr(array_info_dict, array, "NumPy array")
+        elif is_data_io:
+            array_info_dict = get_basic_array_info(array.data)
+            repr_html = generate_array_html_repr(array_info_dict, array.data, "DataIO")
+        elif it_was_read_with_io:  # The backend handles the representation
+            read_io = self.get_read_io()  # Note that sometimes numpy array have a read_io attribute
```

Comment:
In which cases do numpy arrays have a read_io attribute?

Reply:
Starting frames in an `ImageSeries` are numpy even if the object was read with NWBHDF5IO. See here for the previous discussion:

https://github.com/hdmf-dev/hdmf/pull/1201

Reply:
> Starting frames in an `ImageSeries` are numpy even if the object was read with NWBHDF5IO.

Makes sense. This is because `starting_frame` is an Attribute in HDF5 on `external_file` rather and not at Dataset. Attributes are read in while Datasets are loaded lazily. 

Just to clarify, the numpy array itself does not have read_io, but the Container object that owns it. This is similar to the case I mentioned below. The Container has a read_io but the dataset we are looking at has been modified (instead of being the backend dataset). In the case of `starting_frame` the "modification" is that it was read into memory.

Reply:
Oh, yes, I did not realize that the `get_io` method goes to the parent of the object though.

https://github.com/hdmf-dev/hdmf/blob/ea6504f6c49e78d7b3d0effccef2bd1fe19520ab/src/hdmf/container.py#L344-L360

From that perspective the comment is wrong, should be updated.

Reply:
I made an attempt to improve the comments.